File: ride_sharing_processor.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"os"
	"sync"
	"time"
)

type Task struct {
	ID      string
	Payload string
}

func main() {
	numWorkers := 4
	numTasks := 20
	outFile := "results_go.txt"

	log.SetPrefix("[ride-go] ")
	log.SetFlags(log.LstdFlags | log.Lmicroseconds)

	// channels
	taskCh := make(chan Task)
	resultCh := make(chan string)
	var wg sync.WaitGroup

	// writer goroutine: collects results and writes to file at end
	var results []string
	var writerErr error
	var writerWg sync.WaitGroup
	writerWg.Add(1)
	go func() {
		defer writerWg.Done()
		for r := range resultCh {
			results = append(results, r)
		}
		// write to file
		f, err := os.Create(outFile)
		if err != nil {
			writerErr = fmt.Errorf("failed to create output file: %w", err)
			log.Println(writerErr)
			return
		}
		defer f.Close()
		for _, line := range results {
			_, err := f.WriteString(line + "\n")
			if err != nil {
				writerErr = fmt.Errorf("failed to write to file: %w", err)
				log.Println(writerErr)
				return
			}
		}
		log.Println("Writer: results written to", outFile)
	}()

	// start workers
	for w := 1; w <= numWorkers; w++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			log.Printf("Worker %d started\n", workerID)
			for task := range taskCh {
				// process
				res, err := processTask(workerID, task)
				if err != nil {
					log.Printf("Worker %d error processing %s: %v\n", workerID, task.ID, err)
					continue
				}
				// send result (non-blocking assumption: writer keeps up; if writer slow, this will block)
				resultCh <- res
			}
			log.Printf("Worker %d exiting\n", workerID)
		}(w)
	}

	// produce tasks
	go func() {
		for i := 1; i <= numTasks; i++ {
			taskCh <- Task{ID: fmt.Sprintf("task-%02d", i), Payload: fmt.Sprintf("ride-request-%d", i)}
		}
		close(taskCh) // signal no more tasks
	}()

	// wait for workers to finish then close resultCh
	go func() {
		wg.Wait()
		close(resultCh)
	}()

	// wait for writer to finish
	writerWg.Wait()

	if writerErr != nil {
		log.Println("Finished with writer error:", writerErr)
	} else {
		log.Printf("Processing complete. Total results: %d\n", len(results))
	}
}

// processTask simulates work and may return an error sometimes
func processTask(workerID int, t Task) (string, error) {
	// simulate variable processing time
	delay := time.Duration(100+rand.Intn(400)) * time.Millisecond
	time.Sleep(delay)

	// small chance of simulated error
	if rand.Float32() < 0.05 {
		return "", fmt.Errorf("simulated error for %s", t.ID)
	}

	res := fmt.Sprintf("%s | processedBy=worker-%d | at=%s | payload=%s",
		t.ID, workerID, time.Now().Format(time.RFC3339Nano), t.Payload)
	return res, nil
}

File: RideSharingProcessor.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.time.Instant;
import java.util.List;
import java.util.ArrayList;
import java.util.UUID;
import java.util.concurrent.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * RideSharingProcessor
 * - Uses a BlockingQueue for tasks
 * - Fixed thread pool workers consume tasks
 * - Results collected into a thread-safe list
 * - Writes results to a file at the end
 */
public class RideSharingProcessor {
    private static final Logger logger = Logger.getLogger(RideSharingProcessor.class.getName());

    // Simple task representation
    static class Task {
        final String id;
        final String payload;
        Task(String payload) {
            this.id = UUID.randomUUID().toString();
            this.payload = payload;
        }
        @Override
        public String toString() {
            return String.format("Task[id=%s,payload=%s]", id, payload);
        }
    }

    // Poison pill sentinel to stop workers
    private static final Task POISON = new Task("__POISON__");

    public static void main(String[] args) {
        int numWorkers = 4;
        int numTasks = 20;
        String outFile = "results.txt";

        BlockingQueue<Task> queue = new LinkedBlockingQueue<>();
        // Thread-safe results collection
        ConcurrentLinkedQueue<String> results = new ConcurrentLinkedQueue<>();

        ExecutorService workers = Executors.newFixedThreadPool(numWorkers);

        // Worker Runnable
        class Worker implements Runnable {
            private final int workerId;
            Worker(int id) { this.workerId = id; }
            @Override
            public void run() {
                logger.info(() -> "Worker " + workerId + " started.");
                try {
                    while (true) {
                        Task t = queue.take(); // blocking
                        if (t == POISON) {
                            // re-insert poison so other workers can stop too
                            queue.put(POISON);
                            logger.info(() -> "Worker " + workerId + " received poison and will exit.");
                            break;
                        }
                        try {
                            String res = processTask(workerId, t);
                            results.add(res);
                        } catch (Exception e) {
                            logger.log(Level.SEVERE, "Worker " + workerId + " encountered error processing " + t, e);
                        }
                    }
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    logger.log(Level.WARNING, "Worker " + workerId + " interrupted", ie);
                } finally {
                    logger.info(() -> "Worker " + workerId + " finished.");
                }
            }
        }

        // Start workers
        for (int i = 0; i < numWorkers; i++) {
            workers.submit(new Worker(i + 1));
        }

        // Enqueue tasks
        for (int i = 0; i < numTasks; i++) {
            queue.add(new Task("ride-request-" + (i + 1)));
        }

        // Insert a single poison; workers will re-insert it to let all exit
        queue.add(POISON);

        // Shutdown and await termination
        workers.shutdown();
        try {
            if (!workers.awaitTermination(60, TimeUnit.SECONDS)) {
                logger.warning("Forcing shutdown of worker pool...");
                workers.shutdownNow();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            workers.shutdownNow();
        }

        // Write results to file
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outFile))) {
            for (String r : results) {
                bw.write(r);
                bw.newLine();
            }
            logger.info("Results written to " + outFile);
        } catch (IOException ioe) {
            logger.log(Level.SEVERE, "Failed to write results to file", ioe);
        }

        // Also print a small summary
        logger.info(() -> "Processing complete. Tasks processed: " + results.size());
    }

    // Simulated processing method (could raise an exception)
    private static String processTask(int workerId, Task t) throws Exception {
        // Simulate variable processing time
        long delay = 100 + (long)(Math.random() * 400);
        Thread.sleep(delay);

        // Occasionally simulate an error
        if (Math.random() < 0.05) {
            throw new RuntimeException("Simulated processing error for " + t.id);
        }

        String result = String.format("%s | processedBy=worker-%d | at=%s | payload=%s",
                t.id, workerId, Instant.now().toString(), t.payload);
        logger.fine(() -> "Worker " + workerId + " processed " + t.id);
        return result;
    }
}
